scripts/validateDeploymentUtils.ts

```diff
@@ -0,0 +1,300 @@
+  const searchResult = await searchDirectory(buildPath, findContract);
+  if (!searchResult) {
+    throw new Error("Artifact not found");
+  }
+
+  return JSON.parse(fs.readFileSync(searchResult, "utf-8"));
+}
+
+async function getBytecodeWithLinks(contractName: string): Promise<string> {
+  const artifact = await getArtifact(contractName);
+
+  // convert string to array for inplace replacement
+  const arrayBytecode = artifact.bytecode.split("");
+
+  for (const dependency of Object.values(artifact.linkReferences)) {
+    const dependencyName = Object.keys(dependency)[0];
+    const dependencyPositions = dependency[dependencyName];
+
+    const deploymentFilename = "./deployments/" + hre.network.name + "/" + dependencyName + ".json";
+    // extract deployed dependency address and trim 0x
+    const addr = (await getAddressFromDeployment(deploymentFilename)).substring(2);
+
+    for (const position of dependencyPositions) {
+      // calculate link index. Each byte is 2 symbols + 0x at the start.
+      const idx = 2 + position.start * 2;
+      arrayBytecode.splice(idx, 40, addr);
+    }
+  }
+
+  return arrayBytecode.join("");
+}
+
+async function compileContract(contractName: string) {
+  // Find artifact with our contract and remove it to force recompilation of this contract
+  const findContract = findFile(contractName + ".sol");
+  const buildPath = path.join(__dirname, "../artifacts/contracts/");
+  const searchResult = await searchDirectory(buildPath, findContract);
+  if (searchResult) {
+    fs.rmSync(searchResult, { recursive: true, force: true });
+  }
+
+  execSync("npx hardhat compile", { stdio: "inherit" });
+  console.log(`${contractName} compiled successfully.`);
+}
+
+//Using streaming read cause file can be big
+const getAddressFromDeployment = async (filename: string): Promise<string | null> => {
+  if (fs.lstatSync(filename).isDirectory()) {
+    return null;
+  }
+
+  try {
+    const fileStream = fs.createReadStream(filename);
+    const rl = readline.createInterface({ input: fileStream, crlfDelay: Infinity });
+
+    for await (const line of rl) {
+      const match = line.match(/"address":\s*"(0x[^"]+)"/);
+      if (match) {
+        return match[1];
+      }
+    }
+    return null;
+  } catch (error) {
+    console.error("Error reading file:", error);
+    return null;
+  }
+};
+
+const checkAddressInFile =
+  (address: string) =>
+  async (filename: string): Promise<boolean> => {
+    const deploymentAddress = await getAddressFromDeployment(filename);
+    return deploymentAddress === address;
+  };
+
+const findFile =
+  (searchFile: string) =>
+  (filename: string): Promise<boolean> => {
+    return Promise.resolve(filename.endsWith(searchFile));
+  };
+
+// Search recursively through all files in the `dirPath` and test it with `condition`
+// Returns filename when condition is true
+async function searchDirectory(dirPath: string, condition: (filename: string) => Promise<boolean>): Promise<string> {
```

Comment:
at first glance this function could be sync

Reply:
i would also suggest to use https://www.npmjs.com/package/glob

newer nodejs versions support it natively, but 20 (that we use) doesn't have it yet